src/librbd/ImageCtx.cc

```diff
@@ -519,19 +519,29 @@ librados::IoCtx duplicate_io_ctx(librados::IoCtx& io_ctx) {
   uint64_t ImageCtx::get_image_size(snap_t in_snap_id) const
   {
     ceph_assert(ceph_mutex_is_locked(image_lock));
+
+    uint64_t raw_size;
     if (in_snap_id == CEPH_NOSNAP) {
       if (!resize_reqs.empty() &&
           resize_reqs.front()->shrinking()) {
-        return resize_reqs.front()->get_image_size();
+        raw_size = resize_reqs.front()->get_image_size();
+      } else {
+        raw_size = size;
+      }
+    } else {
+      const SnapInfo *info = get_snap_info(in_snap_id);
+      if (info) {
+        raw_size = info->size;
+      } else {
+        raw_size = 0;
       }
-      return size;
     }
 
-    const SnapInfo *info = get_snap_info(in_snap_id);
-    if (info) {
-      return info->size;
-    }
-    return 0;
+    io::Extents extents = {{raw_size, 0}};
```

Comment:
So you're basically suggesting that instead of changing get_image_size, I should change clip_io to correctly verify extent is valid in case encryption is on?

Reply:
The librbd API IO test works by stat-ing the image to get it size, and then test IO is valid for IMAGE_SIZE-1.
Do you suggest also changing stat to account for crypto?

Reply:
Might just need a new API for effective image size then? QEMU would need to be modified anyway to load LUKS prior to any change in size being apparent. Or the `stat` returns the effective size but internally `clip_io` and the `stat` API switch to a new `get_effective_image_size` helper.